src/drag-drop-context.tsx

```diff
@@ -1,30 +1,135 @@
+
+type SensorActivator<K extends keyof HTMLElementEventMap> = (
+  event: HTMLElementEventMap[K],
+  draggableId: string | number
+) => void;
+type Draggable = {
+  id: string | number;
+  node: HTMLElement;
+  layout: Layout;
+  data: Record<string, any>;
+  transform: Transform;
+};
+type Droppable = {
+  id: string | number;
+  node: HTMLElement;
+  layout: Layout;
+  data: Record<string, any>;
+  transform: Transform;
+};
+type DragEvent = {
+  draggable: Draggable;
+  droppable?: Droppable | null;
+};
+type RecomputeFilter = "all" | "draggable" | "droppable";
+
+interface Sensor {
+  id: string | number;
+  activators: { [K in keyof HTMLElementEventMap]?: SensorActivator<K> };
+}
+interface DragDropState {
+  draggables: Record<string | number, Draggable | undefined | null>;
+  droppables: Record<string | number, Droppable | undefined | null>;
+  sensors: Record<string | number, Sensor | undefined | null>;
+  active: {
+    draggable: string | number | null;
+    droppable: string | number | null;
+    sensor: string | number | null;
+  };
+  previous: {
+    draggable: string | number | null;
+    droppable: string | number | null;
+  };
+  usingDragOverlay: boolean;
+}
+interface DragDropActions {
+  setUsingDragOverlay(value?: boolean): void;
+  addDraggable(draggable: Omit<Draggable, "transform">): void;
+  addDroppable(droppable: Omit<Droppable, "transform">): void;
+  removeDraggable(id: string | number): void;
+  removeDroppable(id: string | number): void;
+  addSensor(sensor: Sensor): void;
+  removeSensor(id: string | number): void;
+  sensorStart(id: string | number): void;
+  sensorEnd(): void;
+  recomputeLayouts(filter?: RecomputeFilter): boolean;
+  detectCollisions(): void;
+  displace(
+    type: "draggables" | "droppables",
+    id: string | number,
+    transform: Transform
+  ): void;
+  activeDraggable(): Draggable | null;
+  activeDroppable(): Droppable | null;
+  activeSensor(): Sensor | null;
+  draggableActivators(
+    draggableId: string | number,
+    asHandlers?: boolean
+  ): Listeners;
+  anyDraggableActive(): boolean;
+  anyDroppableActive(): boolean;
+  anySensorActive(): boolean;
+  dragStart(draggableId: string | number): void;
+  dragMove(transform: Transform): void;
+  dragEnd(): void;
+  onDragStart(handler: DragEventHandler): void;
+  onDragMove(handler: DragEventHandler): void;
+  onDragOver(handler: DragEventHandler): void;
+  onDragEnd(handler: DragEventHandler): void;
+}
+interface DragDropContextProps {
+  onDragStart?: DragEventHandler;
+  onDragMove?: DragEventHandler;
+  onDragOver?: DragEventHandler;
+  onDragEnd?: DragEventHandler;
+  collisionDetectionAlgorithm?(
+    layout: Layout,
+    layouts: Layout[]
+  ): Layout | null;
+}
+
+type DragDropContext = [Store<DragDropState>, DragDropActions];
+type Listeners = Record<
+  string,
+  (event: HTMLElementEventMap[keyof HTMLElementEventMap]) => void
+>;
+type DragEventHandler = (event: DragEvent) => void;
```

Comment:
I found `event` being an object felt a bit unnatural now when using - I'm not sure I consider it an 'event' object in practice. In other words, with the surrounding changes, it feels more natural and expected to write `onDragEnd((draggable, droppable)=>{});` than `onDragEnd(({draggable, droppable})=>{});`

What do you think?

Reply:
Good points. Let's keep it as an 'event' interface for now then and see how it goes.